### PR TITLE
Rewrite C++ document classes as polymorphic value types

### PR DESCRIPTION
**Pull request by wilhuff**
The principle here is that all subclasses of `MaybeDocument` have an internal `Rep` (representation) type that subclasses `MaybeDocument::Rep`. An instance of `MaybeDocument` may have a pointer to any subclass's `Rep`, and an instance of the subclass can be recovered by explicitly copying the subclass type.

This makes it possible for a function to accept a parameter of type `const MaybeDocument&` and this will accept instances of type `Document` or `NoDocument` without any special effort.

It also makes it possible to have a vector of MaybeDocument and have `operator==` just work, since such values are value objects from the point of view of the compiler.

If this approach is acceptable I'll apply it to mutations and FieldValue/ObjectValue.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/model/document.h [hunk 1]
```diff
@@ -59,67 +61,64 @@ std::ostream& operator<<(std::ostream& os, DocumentState state);
...
+           DocumentKey key,
+           SnapshotVersion version,
+           DocumentState document_state,
+           absl::any proto);
+
+  /**
+   * Casts a MaybeDocument to a Document. This is a checked operation that will
+   * assert if the type of the MaybeDocument isn't actually Type::Document.
+   */
+  explicit Document(const MaybeDocument& document);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/document.h]: There's a bunch of places where the `MaybeDocument` could be moved into the `Document`, but given that copying it is cheap, it's probably not worth the trouble.

File: Firestore/core/src/firebase/firestore/model/maybe_document.h [hunk 2]
```diff
@@ -51,61 +54,114 @@ class MaybeDocument {
     UnknownDocument,
   };
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/maybe_document.h]: One very interesting property of this class is how copying derived class' objects "just works" without slicing. Perhaps:
* note that in the class comment?
* add checks to ensure slicing cannot occur in the future? Something like `static_assert(sizeof(*this) == sizeof(MaybeDocument)` added to each derived class?

File: Firestore/core/src/firebase/firestore/model/mutation.h [hunk 3]
```diff
@@ -181,19 +178,20 @@ class Mutation {
    *     only if maybe_doc was nullptr and the mutation would not create a new
    *     document.
    */
-  virtual MaybeDocumentPtr ApplyToLocalView(
-      const MaybeDocumentPtr& maybe_doc,
-      const MaybeDocument* base_doc,
+  virtual absl::optional<MaybeDocument> ApplyToLocalView(
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/mutation.h]: I've taken a stab at this, PTAL. @schmidt-sebastian I've updated these comments based on the confusion Costa expressed. WDYT?

File: Firestore/core/src/firebase/firestore/model/precondition.h [hunk 4]
```diff
@@ -110,12 +111,10 @@ class Precondition {
...
   }
 
-  // For Objective-C++ hash; to be removed after migration.
-  // Do NOT use in C++ code.
-  NSUInteger Hash() const {
-    NSUInteger hash = std::hash<Timestamp>()(update_time_.timestamp());
-    hash = hash * 31 + exists_;
-    hash = hash * 31 + static_cast<NSUInteger>(type_);
+  size_t Hash() const {
+    size_t hash = std::hash<Timestamp>()(update_time_.timestamp());
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/precondition.h]: This is coming in a later PR, so I'd rather avoid conflicts.

File: Firestore/core/src/firebase/firestore/model/unknown_document.cc [hunk 5]
```diff
@@ -16,15 +16,40 @@
...
 namespace model {
 
+class UnknownDocument::Rep : public MaybeDocument::Rep {
+ public:
+  Rep(DocumentKey key, SnapshotVersion version)
+      : MaybeDocument::Rep(Type::UnknownDocument, std::move(key), version) {
+  }
+
+  bool has_pending_writes() const override {
+    return true;
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/unknown_document.cc]: Done. @schmidt-sebastian does this comport with your understanding?

File: Firestore/core/src/firebase/firestore/model/unknown_document.cc [hunk 6]
```diff
@@ -23,20 +23,26 @@ namespace firebase {
...
+
 class UnknownDocument::Rep : public MaybeDocument::Rep {
  public:
   Rep(DocumentKey key, SnapshotVersion version)
       : MaybeDocument::Rep(Type::UnknownDocument, std::move(key), version) {
   }
 
   bool has_pending_writes() const override {
+    // Unknown documents can only exist because of a logical inconsistency
+    // between the server successfully committing a mutation that our local
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/unknown_document.cc]: Nit: `between` is missing the counterpart; consider either tweaking the phrasing to something like "between the server successfully committing a mutation _and_ our local cache _believing it_ should not apply", or just replacing between with something like "of" ("the logical inconsistency of the server...").

File: Firestore/core/src/firebase/firestore/model/document.h (hunk 1 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document.h]: I think [here](https://github.com/firebase/firebase-ios-sdk/pull/3531/files#diff-df205eaa49ddf9029c664b8a454f7cceR364) is an example (I should have mentioned they are all in the follow-up PR).

File: Firestore/core/src/firebase/firestore/model/document.h [hunk 7]
```diff
@@ -64,15 +64,15 @@ class Document : public MaybeDocument {
   Document() = default;
 
   /**
-   * Construct a document. ObjectValue must be passed by rvalue.
+   * Constructs a document.
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/document.h]: Removed.

File: Firestore/core/src/firebase/firestore/model/mutation.h [hunk 8]
```diff
@@ -144,38 +144,52 @@ class Mutation {
...
+   * given MaybeDocument.
    *
-   * @param maybe_doc The document to mutate. The input document can be nullptr
-   *     if the client has no knowledge of the pre-mutation state of the
-   *     document.
+   * Unlike ApplyToRemoteDocument, this method is used before the mutation has
+   * been committed and so it's possible that the mutation is operating on a
+   * locally non-existent document and may produce a non-existent document.
+   *
+   * Note: `maybe_doc` and `base_doc` are similar but not the same. When
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/mutation.h]: I took yours verbatim.

File: Firestore/core/src/firebase/firestore/model/unknown_document.cc (hunk 6 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/unknown_document.cc]: Took the former advice.

**Pull request merged**